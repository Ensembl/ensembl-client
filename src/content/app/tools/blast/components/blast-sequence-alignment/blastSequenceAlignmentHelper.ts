/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const ALIGNMENT_LINE_LENGTH = 60;

type Params = {
  querySequence: string;
  hitSequence: string;
  alignmentLine: string;
  queryStart: number;
  queryEnd: number;
  hitStart: number;
  hitEnd: number;
  alignmentLineLength?: number;
};

export type BlastAlignmentLine = {
  queryLineStart: number; // Position in the query sequence at the start of a line. To be used only for labelling purposes
  queryLineEnd: number; // Position in the query sequence at the end of a line. Excludes mismatches while counting. To be used only for labelling purposes
  hitLineStart: number; // Position in the matched sequence at the start of a line. To be used only for labelling purposes
  hitLineEnd: number; // Position in the matched sequence at the end of a line. Excludes mismatches while counting. To be used only for labelling purposes
  alignmentLineStart: number; // Counter starting from the first character of the first line of the alignment
  alignmentLineEnd: number; // Counter starting from the first character of the first line of the alignment
  queryLine: string; // Query sequence; may contain dashes for missing characters relative to hit sequence
  hitLine: string; // Matched sequence; may contain dashes for missing characters relative to query sequence
  alignmentLine: string; // Vertical line characters representing matches between the query and the target sequence
};

export const createBlastSequenceAlignment = (params: Params) => {
  const {
    querySequence,
    hitSequence,
    alignmentLine,
    queryStart,
    hitStart,
    hitEnd,
    alignmentLineLength = ALIGNMENT_LINE_LENGTH
  } = params;

  let cursor = 0;
  const hitDirection = hitEnd > hitStart ? 1 : -1;
  const alignedLines: BlastAlignmentLine[] = [];

  while (cursor < querySequence.length) {
    const lastAlignedSegment = alignedLines.at(-1);
    const nextQueryStartPosition = lastAlignedSegment?.queryLineEnd
      ? lastAlignedSegment?.queryLineEnd + 1 // start next line from the next nucleotide / amino acid
      : queryStart;
    const nextHitStartPosition = lastAlignedSegment?.hitLineEnd
      ? lastAlignedSegment?.hitLineEnd + hitDirection // start nest line from the next nucleotide / amino acid
      : hitStart;

    const alignedSegment = createAlignedSegment({
      queryStart: nextQueryStartPosition,
      hitStart: nextHitStartPosition,
      hitDirection,
      querySequence,
      hitSequence,
      alignmentLine,
      cursor
    });

    alignedLines.push(alignedSegment);
    cursor += alignmentLineLength;
  }

  return alignedLines;
};

const createAlignedSegment = (params: {
  querySequence: string;
  hitSequence: string;
  alignmentLine: string;
  hitDirection: number; // either 1 or - 1
  queryStart: number;
  hitStart: number;
  alignmentLineLength?: number;
  cursor: number;
}): BlastAlignmentLine => {
  const {
    querySequence,
    hitSequence,
    alignmentLine,
    hitDirection,
    queryStart,
    hitStart,
    alignmentLineLength = ALIGNMENT_LINE_LENGTH,
    cursor
  } = params;

  const startIndex = cursor;
  const endIndex =
    querySequence.length - cursor > alignmentLineLength
      ? cursor + alignmentLineLength - 1
      : querySequence.length - 1;

  const querySequenceSlice = querySequence.slice(startIndex, endIndex + 1);
  const alignmentSlice = alignmentLine.slice(startIndex, endIndex + 1);
  const hitSequenceSlice = hitSequence.slice(startIndex, endIndex + 1);

  const updatedQuerySequenceCounter = calculateEndPosition({
    sequence: querySequence,
    startIndex,
    endIndex,
    startPosition: queryStart,
    strandDirectionMultiplier: 1
  });

  const updatedHitSequenceCounter = calculateEndPosition({
    sequence: hitSequence,
    startIndex,
    endIndex,
    startPosition: hitStart,
    strandDirectionMultiplier: hitDirection
  });

  return {
    queryLineStart: queryStart,
    queryLineEnd: updatedQuerySequenceCounter,
    hitLineStart: hitStart,
    hitLineEnd: updatedHitSequenceCounter,
    alignmentLineStart: cursor + 1,
    alignmentLineEnd: cursor + alignmentLineLength,
    queryLine: querySequenceSlice,
    hitLine: hitSequenceSlice,
    alignmentLine: alignmentSlice
  };
};

const calculateEndPosition = ({
  sequence,
  startIndex,
  endIndex,
  startPosition,
  strandDirectionMultiplier
}: {
  sequence: string;
  startIndex: number;
  endIndex: number;
  startPosition: number;
  strandDirectionMultiplier: number;
}) => {
  let position = startPosition;

  for (let i = startIndex; i < endIndex; i++) {
    if (sequence[i] !== '-') {
      position += 1 * strandDirectionMultiplier;
    }
  }

  return position;
};
