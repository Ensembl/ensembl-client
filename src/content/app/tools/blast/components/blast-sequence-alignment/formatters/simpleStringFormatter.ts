/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { BlastAlignmentLine } from '../blastSequenceAlignmentHelper';

type LineLabelGetter = (position: number) => string;
type LineLabel = string | number | LineLabelGetter;

type Params = {
  alignmentLines: BlastAlignmentLine[];
  queryLineStartLabel?: LineLabel;
  queryLineEndLabel?: LineLabel;
  hitLineStartLabel?: LineLabel;
  hitLineEndLabel?: LineLabel;
};

// QUESTION: should we add numbers to the middle line (containing the ASCI art that connects the query line with the hit line?)

export const simpleStringBlastAlignmentFormatter = (params: Params) => {
  const { alignmentLines } = params;

  let result = '';

  for (const alignmentLine of alignmentLines) {
    result += formatAlignmentLine({
      ...params,
      alignmentLine
    });
    result += '\n\n';
  }

  return result;
};

const formatAlignmentLine = (
  params: Params & { alignmentLine: BlastAlignmentLine }
) => {
  const {
    alignmentLine,
    queryLineStartLabel,
    queryLineEndLabel,
    hitLineStartLabel,
    hitLineEndLabel
  } = params;

  const {
    queryLine,
    hitLine,
    alignmentLine: middleLine,
    queryLineStart,
    queryLineEnd,
    hitLineStart,
    hitLineEnd
  } = alignmentLine;

  let queryLinePrefix = getLabel(queryLineStart, queryLineStartLabel);
  const queryLineSuffix = getLabel(queryLineEnd, queryLineEndLabel);
  let hitLinePrefix = getLabel(hitLineStart, hitLineStartLabel);
  const hitLineSuffix = getLabel(hitLineEnd, hitLineEndLabel);

  const prefixLength = Math.max(queryLinePrefix.length, hitLinePrefix.length);
  queryLinePrefix = queryLinePrefix.padEnd(prefixLength, ' ');
  hitLinePrefix = hitLinePrefix.padEnd(prefixLength, ' ');

  const alignmentPrefix = ''.padStart(prefixLength, ' ');

  return [
    `${queryLinePrefix} ${queryLine} ${queryLineSuffix}`,
    `${alignmentPrefix} ${middleLine}`,
    `${hitLinePrefix} ${hitLine} ${hitLineSuffix}`
  ].join('\n');
};

const getLabel = (position: number, labelOrLabelGetter?: LineLabel): string => {
  if (!labelOrLabelGetter) {
    return `${position}`;
  } else if (labelOrLabelGetter instanceof Function) {
    return labelOrLabelGetter(position);
  } else {
    return `${labelOrLabelGetter}`;
  }
};
