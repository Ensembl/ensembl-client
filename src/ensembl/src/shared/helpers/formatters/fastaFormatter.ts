export const LINE_LENGTH = 60; // line length in Ensembl refget implementations

export const toFasta = (sequenceLabel: string, sequence: string) => {
  const formattedSequence = [];
  formattedSequence.push(`>${sequenceLabel}`);

  let row = '';

  for (let i = 0; i < sequence.length; i++) {
    row += sequence[i];

    const isAtEndOfLine = (i + 1) % LINE_LENGTH === 0;
    if (i === sequence.length - 1 || isAtEndOfLine) {
      formattedSequence.push(row);
      row = '';
    }
  }

  return formattedSequence.join('\n');
};
