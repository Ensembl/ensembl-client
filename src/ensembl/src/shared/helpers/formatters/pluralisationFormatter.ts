type PluralDictionary = {
  [key: string]: string;
};

const pluralDictionary: PluralDictionary = {
  species: 'species'
};

export function pluralise(word: string, count: number) {
  if (count === 1) {
    return word;
  } else {
    if (word in pluralDictionary) {
      return pluralDictionary[word];
    } else {
      return `${word}s`;
    }
  }
}
