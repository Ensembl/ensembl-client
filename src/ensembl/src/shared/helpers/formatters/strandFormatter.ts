import { Strand } from 'src/content/app/entity-viewer/types/strand';

export function getStrandDisplayName(strandCode: Strand) {
  if (strandCode === Strand.FORWARD) {
    return 'forward strand';
  } else {
    return 'reverse strand';
  }
}
