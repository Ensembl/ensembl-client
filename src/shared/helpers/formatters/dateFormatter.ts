/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const getFormattedDate = (date: Date) => {
  const year = date.getUTCFullYear();
  const monthAsNumber = date.getUTCMonth() + 1; // month returned from date.getMonth() is zero-based
  const month = String(monthAsNumber).padStart(2, '0');
  const day = String(date.getUTCDate()).padStart(2, '0');

  return `${year}-${month}-${day}`;
};

export const getFormattedDateTime = (date: Date) => {
  const formattedDate = getFormattedDate(date);
  const hours = String(date.getUTCHours()).padStart(2, '0');
  const minutes = String(date.getUTCMinutes()).padStart(2, '0');

  return `${formattedDate}, ${hours}:${minutes}`;
};
