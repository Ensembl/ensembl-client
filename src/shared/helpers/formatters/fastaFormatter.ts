/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const LINE_LENGTH = 60; // line length in Ensembl refget implementations

type Sequence = {
  header?: string;
  value: string;
};

type Options = {
  lineLength?: number;
};

export const toFasta = (sequence: Sequence, options: Options = {}) => {
  const { header, value: rawSequence } = sequence;
  const { lineLength = LINE_LENGTH } = options;

  const formattedSequence = [];

  if (header) {
    formattedSequence.push(`>${header}`);
  }

  let row = '';

  for (let i = 0; i < rawSequence.length; i++) {
    row += rawSequence[i];

    const isAtEndOfLine = (i + 1) % lineLength === 0;
    if (i === rawSequence.length - 1 || isAtEndOfLine) {
      formattedSequence.push(row);
      row = '';
    }
  }

  return formattedSequence.join('\n');
};
