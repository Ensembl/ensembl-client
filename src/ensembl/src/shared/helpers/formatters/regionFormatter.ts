import { getCommaSeparatedNumber } from 'src/shared/helpers/formatters/numberFormatter';

import { EnsObjectLocation } from 'src/shared/state/ens-object/ensObjectTypes';

export const getFormattedLocation = (location: EnsObjectLocation) => {
  const start = getCommaSeparatedNumber(location.start);
  const end = getCommaSeparatedNumber(location.end);

  return `${location.chromosome}:${start}-${end}`;
};
