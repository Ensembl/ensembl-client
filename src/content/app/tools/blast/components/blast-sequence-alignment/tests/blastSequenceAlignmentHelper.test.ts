/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createBlastSequenceAlignment } from '../blastSequenceAlignmentHelper';
import { simpleStringBlastAlignmentFormatter } from '../formatters/simpleStringFormatter';

import {
  shortNucleotideAlignment,
  shortNucleotideAlignmentOppositeStrand
} from './blast-alignment-fixtures/shortAlignment';
import { longGenomicAlignmentOppositeStrand } from './blast-alignment-fixtures/multiLineAlignment';

describe('createBlastSequenceAlignment', () => {
  describe('short, single-line alignment', () => {
    it('produces correct output for same-strand match', () => {
      const input = shortNucleotideAlignment;
      const [alignment] = createBlastSequenceAlignment(input);

      // The input alignment was short enough not to get split into lines;
      // and as a result, the values returned from the function should be the same as what went in
      expect(alignment.queryLine).toEqual(input.querySequence);
      expect(alignment.hitLine).toEqual(input.hitSequence);
      expect(alignment.alignmentLine).toEqual(input.alignmentLine);
      expect(alignment.queryLineStart).toEqual(input.queryStart);
      expect(alignment.queryLineEnd).toEqual(input.queryEnd);
      expect(alignment.hitLineStart).toEqual(input.hitStart);
      expect(alignment.hitLineEnd).toEqual(input.hitEnd);
    });

    // same as for the previous test, really
    it('produces correct output for opposite-strand match', () => {
      const input = shortNucleotideAlignmentOppositeStrand;
      const [alignment] = createBlastSequenceAlignment(input);

      expect(alignment.queryLine).toEqual(input.querySequence);
      expect(alignment.hitLine).toEqual(input.hitSequence);
      expect(alignment.alignmentLine).toEqual(input.alignmentLine);
      expect(alignment.queryLineStart).toEqual(input.queryStart);
      expect(alignment.queryLineEnd).toEqual(input.queryEnd);
      expect(alignment.hitLineStart).toEqual(input.hitStart);
      expect(alignment.hitLineEnd).toEqual(input.hitEnd);
    });
  });

  describe('long alignments wrapping over multiple lines', () => {
    it('produces correct output', () => {
      const input = longGenomicAlignmentOppositeStrand;
      const alignmentLines = createBlastSequenceAlignment(input);

      // Make sure that after all the line breaking, the numbers indicating the position of the last character
      // in the query sequence and in the hit sequence are still correct
      const firstAlignmentLine = alignmentLines[0];
      const lastAlignmentLine = alignmentLines.at(-1);
      expect(lastAlignmentLine?.queryLineEnd).toBe(
        longGenomicAlignmentOppositeStrand.queryEnd
      );
      expect(lastAlignmentLine?.hitLineEnd).toBe(
        longGenomicAlignmentOppositeStrand.hitEnd
      );
      expect(firstAlignmentLine?.alignmentLineStart).toBe(1);
      expect(lastAlignmentLine?.alignmentLineEnd).toBe(
        input.querySequence.length
      );

      const queryLineStartLabel = (position: number) => `Query ${position}`;
      const hitLineStartLabel = (position: number) => `Subject ${position}`;

      // The easiest way to test the correctness of createBlastSequenceAlignment function for multi-line alignments
      // is to generate a formatted string out of the result produced by it
      const formattedAlignment = simpleStringBlastAlignmentFormatter({
        alignmentLines,
        queryLineStartLabel,
        hitLineStartLabel
      });

      expect(formattedAlignment).toMatchSnapshot();
    });
  });
});
