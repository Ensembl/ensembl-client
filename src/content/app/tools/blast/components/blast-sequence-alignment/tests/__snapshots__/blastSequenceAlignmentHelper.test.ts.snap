// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createBlastSequenceAlignment > long alignments wrapping over multiple lines > produces correct output 1`] = `
"       Query 10 TTCAACAGGAGCAGGAGAACAAAATCCAGTCTCATAGCAGAGGAGAATATAATGTTTACA 69
              1 ||||||||||||||  |||||||||||  ||| | |||||| | |||||  ||||||||| 60
Subject 1095275 TTCAACAGGAGCAGATGAACAAAATCCCATCTTA-AGCAGAAGGGAATACCATGTTTACA 1095217

       Query 70 GCACCTTCCAGAGCCATGAACCAGAGTTTGACTACTTGAAAAGTTTAGAAATAGAAGAAA 129
             61 | ||||| |||| | |||| |||||||||||| |||||||||||||||||| |||||||| 120
Subject 1095216 GGACCTTTCAGAACGATGAGCCAGAGTTTGACCACTTGAAAAGTTTAGAAAGAGAAGAAA 1095157

      Query 130 AGATCAACAAAATTAGGTGGTTACCCC-AGAAAAATGCTGCTCAGTTTTTATTGTCTACC 188
            121 || ||||||||  ||||||||| |||  |||||||| |||||||||    |||||||||| 180
Subject 1095156 AGTTCAACAAAGCTAGGTGGTTTCCCTGAGAAAAATTCTGCTCAGT----ATTGTCTACC 1095101

      Query 189 AATGATAAAACAATAAAATTATGGAAAATCAGTGAAAGGGACAAAAGACCAGAAGGGTAT 248
            181 ||| ||||||||||| |||||||||||||||||||||||||||||||||||||||| ||| 240
Subject 1095100 AATAATAAAACAATACAATTATGGAAAATCAGTGAAAGGGACAAAAGACCAGAAGGATAT 1095041

      Query 249 AACTTGAAAGAGGAGGATGGAAGGTATAGAGATCCTACTACAGTTACTACACTACGAGTG 308
            241 |||||| ||||||| |||||   || ||||||| | || || | |||| ||||  ||||| 300
Subject 1095040 AACTTG-AAGAGGAAGATGGGGTGTGTAGAGATGCCACCACGGCTACTGCACTGTGAGTG 1094982

      Query 309 CCAGTCTTTAGGCCTATGGATCTAATGGTTGAGGCCAGTCCACGAAGAATATTTGCCAAT 368
            301 |||||| |||||||  ||| ||||  ||| |  | || ||||| ||||  || ||||||| 360
Subject 1094981 CCAGTCGTTAGGCCCGTGGGTCTAGGGGTCGCAGTCACTCCACAAAGAGCATCTGCCAAT 1094922

      Query 369 GCTCATACATATCACATCAACTCAATTTCTATTAATAGTGATTATGAAACATATTTATCT 428
            361 ||||||||   ||||| |||||||||| || ||||||| |||| |||||||||||||||| 420
Subject 1094921 GCTCATACTGGTCACACCAACTCAATTCCTGTTAATAGAGATTTTGAAACATATTTATCT 1094862

      Query 429 GCAGATGATTTGCGGATTAA-------TCTTTGGCATCTGGAAATTACAGACAGGAGTTT 481
            421 |||| |||||||| ||| ||       |||||||||||||||||||| |||||||||||| 480
Subject 1094861 GCAGGTGATTTGCAGATCAAAATCAAGTCTTTGGCATCTGGAAATTATAGACAGGAGTTT 1094802

      Query 482 TAACATTGTGGATATCAAGCCTGCCAATATGGAAGAGCTAACAGAGGTGATTACAGCAGC 541
            481 ||| ||||||||||||||||||| ||||||||||||||||||  |||||||| |||| || 540
Subject 1094801 TAATATTGTGGATATCAAGCCTGACAATATGGAAGAGCTAAC--AGGTGATTTCAGCCGC 1094744

      Query 542 AGAATTTCATCCAAACAGCTGTAACACATTTGTATACAGCAGCAGTAAAGGAACTATTCG 601
            541 |||||| ||||  ||||||| |||||||| |||||  ||||||||||||||||||||| | 600
Subject 1094743 AGAATTCCATCTGAACAGCTATAACACATCTGTATCTAGCAGCAGTAAAGGAACTATTTG 1094684

      Query 602 GCTATGTGACATGAGGGCATCTGCCCTCTGTGATAGACATTCTAAATTGTTTGAAGAACC 661
            601    |||||||||||| || ||||||||||||||||| || |||| |     ||||||||| 660
Subject 1094683 CTCATGTGACATGAGAGCGTCTGCCCTCTGTGATAGGCA-TCTACA----CTGAAGAACC 1094629

      Query 662 TGAAGATCCCAGTAACAGGTCATTTTTTTCCGAAATCATCTCCTCTATTTCGGATGTAAA 721
            661 ||| ||| |||||||  |||||||||| || ||||||||||     ||||| || ||||| 720
Subject 1094628 TGATGATTCCAGTAATGGGTCATTTTTGTCTGAAATCATCT----AATTTCTGACGTAAA 1094573

      Query 722 ATTCAGCCATAGTGGTCGATATATGATGACTAGAGACTATTTGTCAGTCAAAATTTGGGA 781
            721 ||||||||||||||||| | ||||||||| ||||||| |||||| |||||||| |||| | 780
Subject 1094572 ATTCAGCCATAGTGGTC-AAATATGATGA-TAGAGACCATTTGTTAGTCAAAAGTTGGAA 1094515

      Query 782 CTTAAATATGGAAAAC 797
            781 ||| |||||||||||| 796
Subject 1094514 CTTCAATATGGAAAAC 1094499

"
`;
